### PR TITLE
Add new changes to the PubSub package

### PR DESCRIPTION
**Pull request by Nuvindu**
Fixes: [#2907](https://github.com/ballerina-platform/ballerina-standard-library/issues/2907)

### CODE REVIEW

File: ballerina/pubsub.bal [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        if self.isClosed is true {
+            return error Error("Unsubscribing is not allowed in a closed PubSub.");
+        }
+        pipe:Pipe[] pipes = <pipe:Pipe[]>self.events[eventName];
+        pipe:Pipe[] modifiedPipes = [];
+        foreach pipe:Pipe element in pipes {
+            if pipe !== element {
+                modifiedPipes.push(element);
+            }
+        }
```
Comment (ThisaruGuruge) [inline ballerina/pubsub.bal:80]: Wouldn't it be easier and faster to just remove the pipe from the existing pipes?

```suggestion
        pipe:Pipe[] pipes = <pipe:Pipe[]>self.events[eventName];
        int i = 0;
        while i < pipes.length() {
            if pipe === element {
                pipes.remove(i);
                break;
            }
            i += 1;
        }
```

File: ballerina/pubsub.bal [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        pipe:Pipe[]? pipes = self.events[eventName];
+        if pipes == () {
+            self.events[eventName] = [pipe];
+        } else {
+            pipes.push(pipe);
+            self.events[eventName] = pipes;
+        }
+    }
+
+    # Adding a topic to the PubSub
```
Comment (ThisaruGuruge) [inline ballerina/pubsub.bal:97]: We have to use consistent wording here. It should be either `topic` or `event`. Using them interchangeably will make it ambiguous.

File: ballerina/pubsub.bal [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        if self.events.hasKey(eventName) is true {
+            return error Error("Event name already exists.");
+        }
+        self.events[eventName] = [];
+    }
+
+    # Close the PubSub gracefully. Waits for some period until all the pipes in the PubSub is gracefully closed.
+    # + return - Return (), if the PubSub is successfully closed. Otherwise returns an error.
+    public function gracefulShutdown() returns error? {
```
Comment (ThisaruGuruge) [inline ballerina/pubsub.bal:113]: Shell we update the return type to `Error`? Check other places as well.

File: ballerina/pubsub.bal [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        self.isClosed = true;
+        foreach pipe:Pipe[] pipes in self.events {
+            foreach pipe:Pipe pipe in pipes {
+                error? gracefulClose = pipe.gracefulClose();
+                if gracefulClose is error {
+                    return gracefulClose;
+                }
```
Comment (ThisaruGuruge) [inline ballerina/pubsub.bal:120]: If we use just `error` here, the following would be sufficient. 

```suggestion
                check pipe.gracefulClose();
```

But if we are to use `Error?` as the return type, the following should be the case:

```ballerina
                error? gracefulClose = pipe.gracefulClose();
                if gracefulClose is error {
                    return new Error("Failed to shut down the pubsub", gracefulClose);
                }
```

File: ballerina/pubsub.bal [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            }
+        }
+        self.events.removeAll();
+    }
+
+    # Close all the Pipes in the PubSub instantly using the immediate close method.
+    public function forceShutdown() {
```
Comment (ThisaruGuruge) [inline ballerina/pubsub.bal:127]: Shouldn't this return an error in case of a failure?

File: build-config/resources/Ballerina.toml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+[package]
+org = "nuvindu_dias"
+name = "pubsub"
+version = "0.1.0"
```
Comment (ThisaruGuruge) [inline build-config/resources/Ballerina.toml:4]: This should have a placeholder to replace the version at the build time.
```suggestion
version = "@toml.version@"
```

File: build-config/resources/Ballerina.toml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+keywords = ["pubsub"]
+repository = "https://github.com/Nuvindu/module-pubsub"
+license = ["Apache-2.0"]
+distribution = "2201.0.1"
+
+[[dependency]]
+org = "nuvindu_dias"
+name = "pipe"
+version = "0.1.0"
+repository = "local"
```
Comment (ThisaruGuruge) [inline build-config/resources/Ballerina.toml:15]:
```suggestion
repository = "local"

```

File: docs/proposals/PubSub.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# Proposal: Pub/Sub Package for Ballerina
+
```
Comment (ThisaruGuruge) [inline docs/proposals/PubSub.md:2]: Let's add the owners and other information about the proposal here.

File: examples/news-alerts/.devcontainer.json [hunk 9]
```diff
@@ -0,0 +1,4 @@
+{
+    "image": "ballerina/ballerina-devcontainer:2201.0.1",
+    "extensions": ["WSO2.ballerina"],
+}
```
Comment (ThisaruGuruge) [inline examples/news-alerts/.devcontainer.json:4]: We can delete this file.

File: examples/news-alerts/Ballerina.toml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+org = "nuvindu_dias"
+name = "pipe"
+version = "0.1.0"
+repository = "local"
+
+[[dependency]]
+org = "nuvindu_dias"
+name = "pubsub"
+version = "0.1.0"
+repository = "local"
```
Comment (ThisaruGuruge) [inline examples/news-alerts/Ballerina.toml:20]: We have to remove these local dependencies.

File: examples/news-alerts/Dependencies.toml [hunk 11]
```diff
@@ -0,0 +1,112 @@
+# AUTO-GENERATED FILE. DO NOT MODIFY.
```
Comment (ThisaruGuruge) [inline examples/news-alerts/Dependencies.toml:1]: Let's remove the Dependencies.toml file for the examples and add it to the gitignore file.

File: gradle.properties [hunk 12]
```diff
@@ -0,0 +1,15 @@
+org.gradle.caching=true
+group=io.ballerina.stdlib
+version=1.3.0-SNAPSHOT
```
Comment (ThisaruGuruge) [inline gradle.properties:3]: Let's use 0.1.0 as the version. (Use the same for the Pipe package as well).

File: settings.gradle [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    }
+}
+
+plugins {
+    id "com.gradle.enterprise" version "3.2"
+}
+
+include ':checkstyle'
+include ':pubsub-ballerina'
+include ':examples'
```
Comment (ThisaruGuruge) [inline settings.gradle:25]:
```suggestion
include ':pubsub-examples'
```

File: native/src/main/java/org/nuvindu/pubsub/PubSub.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+ */
+public class PubSub {
+
+    public static Object subscribe(Environment environment, BObject pubsub, BString topicName, int limit,
+                                   BDecimal timeout, BTypedesc typeParam) {
+        if ((pubsub.get(StringUtils.fromString(Constants.IS_CLOSED))).equals(true)) {
+            return createError("Users cannot subscribe to a closed PubSub.");
+        }
+        BObject defaultPipe = pubsub.getObjectValue(StringUtils.fromString(PIPE));
+        Module module = new Module(ORGANIZATION, PIPE, defaultPipe.getType().getPackage().getMajorVersion());
```
Comment (ThisaruGuruge) [inline native/src/main/java/org/nuvindu/pubsub/PubSub.java:39]: In this case, we don't need to create constants for the organization name, package name, etc. We can retrieve them using the `Package` object (`defaultPipe.getType().getPackage()`). Even better, you might be able to get the `Module` instance directly.

**Review by ThisaruGuruge — COMMENTED** [review]
Let's check for missing license headers and add them.

File: ballerina/Module.md [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+pubsub:PubSub pubsub = new(autoCreateTopics = true);
+```
+
+#### Publish
+
+Events can be published into a topic using this method. Once an event is published, it will be broadcast to all the subscribers of that topic.
+
+```ballerina
+pubsub:Error publish = pubsub.publish(topicName = "topic");
```
[the remark this answers is not in the document]
Comment (Nuvindu) [reply ballerina/Module.md:24]: Sure

**Review by ThisaruGuruge — APPROVED** [review]

**Pull request merged**